bootstrap/src/basis/file.ml

```diff
@@ -130,49 +130,76 @@ let close t =
 let close_hlt t =
   Close.(submit t |> complete_hlt)
 
-let read_n = 1024L
+module Read = struct
+  type file = t
+  type inner = uns
+  type t = {
+    inner: inner;
+    buffer: Bytes.Slice.t;
+  }
+
+  let default_n = 1024L
+
+  external submit_inner: uns -> file -> inner = "hm_basis_file_read_submit_inner"
+
+  let submit ?n ?buffer file =
+    let n, buffer = begin
+      match n with
+      | None -> begin
+          match buffer with
+          | None -> default_n, Bytes.Slice.init (Array.init (0L =:< default_n)
+            ~f:(fun _ -> Byte.kv 0L))
+          | Some buffer -> Bytes.Slice.length buffer, buffer
+        end
+      | Some n -> begin
+          match buffer with
+          | None -> n, Bytes.Slice.init (Array.init (0L =:< n) ~f:(fun _ -> Byte.kv 0L))
+          | Some buffer -> (Uns.min n (Bytes.Slice.length buffer)), buffer
+        end
+    end in
+    let inner = finalize_user_data (submit_inner n file) in
+    { inner; buffer }
 
-(* external read_inner: !&bytes -> t >os-> int *)
-external read_inner: Stdlib.Bytes.t -> t -> sint = "hm_basis_file_read_inner"
+  external complete_inner: Stdlib.Bytes.t -> inner -> sint =
+    "hm_basis_file_read_complete_inner"
 
-let read_impl buffer t =
-  let base = Bytes.(Cursor.index (Slice.base buffer)) in
-  let bytes = Stdlib.Bytes.create (Int64.to_int (Bytes.Slice.length buffer)) in
-  let value = read_inner bytes t in
-  match Sint.(value < kv 0L) with
-  | true -> Error (Uns.bits_of_sint Sint.(neg value))
-  | false -> begin
-      let range = (base =:< (base + (Uns.bits_of_sint value))) in
-      let container = Bytes.Slice.container buffer in
-      Range.iter range ~f:(fun i ->
-        Array.set_inplace i (U8.of_char (Stdlib.Bytes.get bytes (Int64.to_int (i - base))))
-          container
-      );
-      Ok (Bytes.Slice.init ~range (Bytes.Slice.container buffer))
-    end
+  let complete t =
+    let base = Bytes.(Cursor.index (Slice.base t.buffer)) in
+    let bytes = Stdlib.Bytes.create (Int64.to_int (Bytes.Slice.length t.buffer)) in
+    let value = complete_inner bytes t.inner in
+    match Sint.(value < kv 0L) with
+    | true -> Error (Uns.bits_of_sint Sint.(neg value))
+    | false -> begin
+        let range = (base =:< (base + (Uns.bits_of_sint value))) in
+        let container = Bytes.Slice.container t.buffer in
+        Range.iter range ~f:(fun i ->
```

Comment:
I think `Range.iter` can just be `iter` here.

Reply:
Tried it. Doesn't seem to work.

Reply:
Hmm, yeah, `=:<` is in `Rudiments`, but `iter` isn't. At one point I included all of `Range` in `Rudiments`, but that was a mess.